### PR TITLE
AB#41007 Refactor Search

### PR DESCRIPTION
**Pull request by jamesgriff**
* Split out these bits from AdminSearchService
  * SearchRepository
  * SearchCacheUpdaterService
  * 2 cached objects (for Organisations and Users)
* Keep AdminSearchService for the admin-specific search logic
* Make the Admin Search able to search by acronyms (e.g. "bbc" for "British Broadcasting Corporation Limited"

### CODE REVIEW

File: GenderPayGap.WebUI/Search/CachedObjects/SearchReadyValue.cs [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+        public SearchReadyValue(string originalValue)
+        {
+            OriginalValue = originalValue;
+            LowercaseWords = SplitValueIntoWords(originalValue);
+            Acronym = MakeAcronymFromWords(LowercaseWords);
+        }
+
+        private static List<string> SplitValueIntoWords(string originalValue)
+        {
+            Match match = WordSplittingRegex.Regex.Match(originalValue);
```
Comment (raamSoftwire) [inline GenderPayGap.WebUI/Search/CachedObjects/SearchReadyValue.cs]: Will this regex also remove any punctuation (eg. co-op => coop etc)?

File: GenderPayGap.WebUI/Search/WordSplittingRegex.cs [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+                if (regex != null)
+                {
+                    return regex;
+                }
+
+                // The "language=regexp" comments below tell Resharper to do regex syntax highlighting
+                // "CG" means "character group" - i.e. [A-Z] in regex
+
+                //language=regexp
+                string wordCG = "[a-zA-Z0-9'-]";
```
Comment (raamSoftwire) [inline GenderPayGap.WebUI/Search/WordSplittingRegex.cs]: Is a hyphen here counted as a word character then? So `co-op` would remain as `co-op`?

Comment (raamSoftwire) [reply GenderPayGap.WebUI/Search/WordSplittingRegex.cs]: It's hard to tell because a hyphen is used for `a-z` too

File: GenderPayGap.WebUI/Search/SearchRepository.cs [hunk 3]
```diff
@@ -19,6 +19,8 @@ public class SearchRepository
         private static List<SearchCachedUser> cachedUsers;
         private static DateTime cacheLastUpdated = DateTime.MinValue;
 
+
+        #region Load data into cache
```
Comment (raamSoftwire) [inline GenderPayGap.WebUI/Search/SearchRepository.cs]: I'm still unsure about the use of `#regions`. It is a code smell and one that I was hoping to slowly phase out of the codebase. But I can see that it does a good job here of grouping things together and the file size is still relatively small.

I guess this is just a low-level grumble...

File: GenderPayGap.WebUI/Search/SearchRepository.cs [hunk 4]
```diff
@@ -82,6 +86,8 @@ private List<string> ExtractSearchTermsFromQuery(string query)
                 .ToList();
         }
 
+
+        #region Search Organisations
         private List<AdminSearchResultOrganisationViewModel> SearchOrganisations(string query, List<string> searchTerms)
```
Comment (raamSoftwire) [inline GenderPayGap.WebUI/Search/SearchRepository.cs]: The `#regions` do a nice job of providing a narrative, but I think the same is achieved just with the method names here. Which makes the regions feel a bit redundant

File: GenderPayGap.WebUI/Search/CachedObjects/SearchReadyValue.cs (hunk 1 above)
Comment (jamesgriff) [reply GenderPayGap.WebUI/Search/CachedObjects/SearchReadyValue.cs]: Good point, I missed that case.
Just about to push a change that will include this.

File: GenderPayGap.WebUI/Search/WordSplittingRegex.cs (hunk 2 above)
Comment (jamesgriff) [reply GenderPayGap.WebUI/Search/WordSplittingRegex.cs]: Regex patterns are always a bit hard to parse

`[]` means a set of characters that includes `a-z`, `A-Z`, `0-9`, `'` and `-`
So `'` and `-` are part of the word.

The change I'm about to push removes punctuation *after* splitting into words.
So, `co-operative group` becomes [`cooperative`, `group`] rather than [`co`, `operative`, `group`]
